File: frontend/js/views/ev_ebitda.js
// Enterprise value + EV/EBITDA — capital-structure-neutral valuation multiple,
// with EV/Sales and EBITDA margin, via /calc/ev-ebitda. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 0, maximumFractionDigits: 0 });
const mult = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '×');
const pctv = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 1 }) + '%');

export async function renderEvEbitda(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.evebitda.h1.title">// EV / EBITDA</span></h1>
        <p class="muted small" data-i18n="view.evebitda.hint.intro">
            Enterprise value is what it costs to buy the whole business — equity plus the debt you
            assume, less the cash you get. EV/EBITDA is the capital-structure-neutral valuation
            multiple: unlike P/E it ignores how the firm is financed. Add revenue for EV/Sales and
            the EBITDA margin. All figures in the same units (millions or dollars). Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.evebitda.h2.inputs">The company</h2>
            <form id="evebitda-form" class="inline-form">
                <label><span data-i18n="view.evebitda.label.mcap">Market cap</span>
                    <input type="number" step="0.01" min="0" name="market_cap_usd" value="1000" required></label>
                <label><span data-i18n="view.evebitda.label.debt">Total debt</span>
                    <input type="number" step="0.01" min="0" name="total_debt_usd" value="300" required></label>
                <label><span data-i18n="view.evebitda.label.cash">Cash & equivalents</span>
                    <input type="number" step="0.01" min="0" name="cash_usd" value="100" required></label>
                <label><span data-i18n="view.evebitda.label.ebitda">EBITDA</span>
                    <input type="number" step="0.01" name="ebitda_usd" value="150" required></label>
                <fieldset class="inline-fieldset">
                    <legend data-i18n="view.evebitda.legend.optional">Optional</legend>
                    <label><span data-i18n="view.evebitda.label.preferred">Preferred equity</span>
                        <input type="number" step="0.01" min="0" name="preferred_equity_usd" value="0"></label>
                    <label><span data-i18n="view.evebitda.label.minority">Minority interest</span>
                        <input type="number" step="0.01" min="0" name="minority_interest_usd" value="0"></label>
                    <label><span data-i18n="view.evebitda.label.revenue">Revenue</span>
                        <input type="number" step="0.01" min="0" name="revenue_usd" value="0"></label>
                </fieldset>
            </form>
        </div>
        <div id="evebitda-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#evebitda-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            market_cap_usd: Number(fd.get('market_cap_usd')) || 0,
            total_debt_usd: Number(fd.get('total_debt_usd')) || 0,
            cash_usd: Number(fd.get('cash_usd')) || 0,
            ebitda_usd: Number(fd.get('ebitda_usd')) || 0,
            preferred_equity_usd: Number(fd.get('preferred_equity_usd')) || 0,
            minority_interest_usd: Number(fd.get('minority_interest_usd')) || 0,
            revenue_usd: Number(fd.get('revenue_usd')) || 0,
        };
        try {
            const r = await api.calcEvEbitda(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.evebitda.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#evebitda-result');
    const salesRows = r.ev_sales == null ? '' : `
        <tr><td data-i18n="view.evebitda.row.evsales">EV / Sales</td><td>${mult(r.ev_sales)}</td></tr>
        <tr><td data-i18n="view.evebitda.row.margin">EBITDA margin</td><td>${pctv(r.ebitda_margin_pct)}</td></tr>`;
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.evebitda.h2.result">The multiple</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.evebitda.card.ev">Enterprise value</div>
                    <div class="value pos">${money(r.enterprise_value_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.evebitda.card.multiple">EV / EBITDA</div>
                    <div class="value">${mult(r.ev_ebitda)}</div></div>
                <div class="card"><div class="label" data-i18n="view.evebitda.card.netdebt">Net debt</div>
                    <div class="value">${money(r.net_debt_usd)}</div></div>
            </div>
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.evebitda.row.ev">Enterprise value</td><td>${money(r.enterprise_value_usd)}</td></tr>
                    <tr><td data-i18n="view.evebitda.row.netdebt">Net debt (debt − cash)</td><td>${money(r.net_debt_usd)}</td></tr>
                    ${salesRows}
                    <tr class="emph"><td data-i18n="view.evebitda.row.multiple">EV / EBITDA</td><td>${mult(r.ev_ebitda)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/ev_ebitda.rs
//! Enterprise value and the EV/EBITDA multiple.
//!
//! Enterprise value is what it would cost to buy the whole business — equity
//! plus the debt you assume, less the cash you get:
//!
//! ```text
//! EV = market cap + total debt + preferred equity + minority interest − cash
//! ```
//!
//! EV/EBITDA is the capital-structure-neutral valuation multiple (unlike P/E it
//! ignores how the firm is financed), and EV/Sales and the EBITDA margin round
//! out the picture when revenue is supplied.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct EvEbitdaInput {
    /// Equity market capitalization.
    pub market_cap_usd: f64,
    /// Total debt (short- and long-term).
    pub total_debt_usd: f64,
    /// Cash and cash equivalents.
    pub cash_usd: f64,
    /// EBITDA (earnings before interest, taxes, depreciation, amortization).
    pub ebitda_usd: f64,
    /// Preferred equity, if any.
    #[serde(default)]
    pub preferred_equity_usd: f64,
    /// Minority (non-controlling) interest, if any.
    #[serde(default)]
    pub minority_interest_usd: f64,
    /// Revenue, for the EV/Sales multiple and EBITDA margin. Optional.
    #[serde(default)]
    pub revenue_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct EvEbitdaResult {
    /// market cap + debt + preferred + minority − cash.
    pub enterprise_value_usd: f64,
    /// total debt − cash.
    pub net_debt_usd: f64,
    /// EV / EBITDA; `None` if EBITDA is not positive.
    pub ev_ebitda: Option<f64>,
    /// EV / revenue; `None` if revenue is not positive.
    pub ev_sales: Option<f64>,
    /// EBITDA / revenue, percent; `None` if revenue is not positive.
    pub ebitda_margin_pct: Option<f64>,
}

pub fn analyze(input: &EvEbitdaInput) -> EvEbitdaResult {
    let ev = input.market_cap_usd + input.total_debt_usd + input.preferred_equity_usd
        + input.minority_interest_usd
        - input.cash_usd;

    let ev_ebitda = if input.ebitda_usd > 0.0 {
        Some(ev / input.ebitda_usd)
    } else {
        None
    };
    let ev_sales = if input.revenue_usd > 0.0 {
        Some(ev / input.revenue_usd)
    } else {
        None
    };
    let ebitda_margin_pct = if input.revenue_usd > 0.0 {
        Some(input.ebitda_usd / input.revenue_usd * 100.0)
    } else {
        None
    };

    EvEbitdaResult {
        enterprise_value_usd: ev,
        net_debt_usd: input.total_debt_usd - input.cash_usd,
        ev_ebitda,
        ev_sales,
        ebitda_margin_pct,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }

    fn run(mcap: f64, debt: f64, cash: f64, ebitda: f64) -> EvEbitdaResult {
        analyze(&EvEbitdaInput {
            market_cap_usd: mcap,
            total_debt_usd: debt,
            cash_usd: cash,
            ebitda_usd: ebitda,
            preferred_equity_usd: 0.0,
            minority_interest_usd: 0.0,
            revenue_usd: 0.0,
        })
    }

    #[test]
    fn enterprise_value_basic() {
        // 1000 + 300 − 100 = 1200.
        let r = run(1000.0, 300.0, 100.0, 150.0);
        assert!(close(r.enterprise_value_usd, 1200.0));
    }

    #[test]
    fn enterprise_value_with_preferred_and_minority() {
        let r = analyze(&EvEbitdaInput {
            market_cap_usd: 1000.0,
            total_debt_usd: 300.0,
            cash_usd: 100.0,
            ebitda_usd: 150.0,
            preferred_equity_usd: 50.0,
            minority_interest_usd: 20.0,
            revenue_usd: 0.0,
        });
        assert!(close(r.enterprise_value_usd, 1270.0));
    }

    #[test]
    fn ev_ebitda_multiple() {
        // 1200 / 150 = 8.0.
        let r = run(1000.0, 300.0, 100.0, 150.0);
        assert!(close(r.ev_ebitda.unwrap(), 8.0));
    }

    #[test]
    fn net_debt() {
        let r = run(1000.0, 300.0, 100.0, 150.0);
        assert!(close(r.net_debt_usd, 200.0));
    }

    #[test]
    fn negative_net_debt_when_cash_rich() {
        let r = run(1000.0, 100.0, 400.0, 150.0);
        assert!(close(r.net_debt_usd, -300.0));
        // Cash-rich firm: EV below market cap.
        assert!(r.enterprise_value_usd < 1000.0);
    }

    #[test]
    fn negative_ebitda_has_no_multiple() {
        let r = run(1000.0, 300.0, 100.0, -50.0);
        assert!(r.ev_ebitda.is_none());
    }

    #[test]
    fn ev_sales_and_margin_with_revenue() {
        let r = analyze(&EvEbitdaInput {
            market_cap_usd: 1000.0,
            total_debt_usd: 300.0,
            cash_usd: 100.0,
            ebitda_usd: 150.0,
            preferred_equity_usd: 0.0,
            minority_interest_usd: 0.0,
            revenue_usd: 600.0,
        });
        // EV 1200 / 600 = 2.0; margin 150/600 = 25%.
        assert!(close(r.ev_sales.unwrap(), 2.0));
        assert!(close(r.ebitda_margin_pct.unwrap(), 25.0));
    }

    #[test]
    fn no_revenue_no_sales_metrics() {
        let r = run(1000.0, 300.0, 100.0, 150.0);
        assert!(r.ev_sales.is_none());
        assert!(r.ebitda_margin_pct.is_none());
    }
}
